crates/vm/levm/src/vm.rs

```diff
@@ -346,32 +349,93 @@ impl VM {
-        // Note: currently working with copies
-        let call_frame = self
-            .call_frames
-            .last()
-            .ok_or(VMError::Internal(
-                InternalError::CouldNotAccessLastCallframe,
-            ))?
-            .clone();
+    fn add_intrinsic_gas(&mut self, initial_call_frame: &mut CallFrame) -> Result<(), VMError> {
+        // Intrinsic gas is the gas consumed by the transaction before the execution of the opcodes. Section 6.2 in the Yellow Paper.
+
+        // Intrinsic Gas = Calldata cost + Create cost + Base cost + Access list cost
+        let mut intrinsic_gas: U256 = U256::zero();
+
+        // Calldata Cost
+        // 4 gas for each zero byte in the transaction data 16 gas for each non-zero byte in the transaction.
+        let calldata_cost =
+            gas_cost::tx_calldata(&initial_call_frame.calldata).map_err(VMError::OutOfGas)?;
+
+        intrinsic_gas = intrinsic_gas
+            .checked_add(calldata_cost)
+            .ok_or(VMError::OutOfGas(OutOfGasError::ConsumedGasOverflow))?;
+
+        // Base Cost
+        intrinsic_gas = intrinsic_gas
+            .checked_add(TX_BASE_COST)
+            .ok_or(VMError::OutOfGas(OutOfGasError::ConsumedGasOverflow))?;
 
-        self.decrement_account_nonce(call_frame.msg_sender)?;
+        // Create Cost
+        if self.is_create() {
+            intrinsic_gas = intrinsic_gas
+                .checked_add(CREATE_BASE_COST)
+                .ok_or(VMError::OutOfGas(OutOfGasError::ConsumedGasOverflow))?;
 
-        let new_contract_address = call_frame.to;
-        if cache::remove_account(&mut self.cache, &new_contract_address).is_none() {
-            return Err(VMError::AddressDoesNotMatchAnAccount); // Should not be this error
+            let number_of_words: u64 = initial_call_frame
+                .calldata
+                .chunks(WORD_SIZE)
+                .len()
+                .try_into()
+                .map_err(|_| VMError::Internal(InternalError::ConversionError))?;
+
+            intrinsic_gas = intrinsic_gas
+                .checked_add(
+                    U256::from(number_of_words)
+                        .checked_mul(U256::from(2))
+                        .ok_or(VMError::OutOfGas(OutOfGasError::ConsumedGasOverflow))?,
+                )
+                .ok_or(VMError::OutOfGas(OutOfGasError::ConsumedGasOverflow))?;
         }
 
+        // Access List Cost
+        // TODO: Implement access list cost.
+
+        self.increase_consumed_gas(initial_call_frame, intrinsic_gas)
+            .map_err(|_| VMError::TxValidation(TxValidationError::IntrinsicGasTooLow))?;
+
         Ok(())
     }
 
+    /// To get the maximum fee per gas that the user is willing to pay, independently of the actual gas price
+    /// For legacy transactions the max fee per gas is the gas price
+    fn max_fee_per_gas_or_gasprice(&self) -> U256 {
+        self.env.tx_max_fee_per_gas.unwrap_or(self.env.gas_price)
```

Comment:
Is this ok? Generally, the gas price is either the transaction type 0 or 1 gas price or the transaction type 2 or 3 `max_fee_per_gas` but not the reciprocal.

Reply:
It is indeed necessary for validations only. Not for execution.
For example there's a validation named `INSUFFICIENT_MAX_FEE_PER_GAS` that checks if the max fee per gas is less than the base fee per gas. The problem is that we also have to return that error for type 0 and 1 transactions when gas price is less than base fee per gas.
We also need it the most for the calculation of the **up-front cost**, where `Tp` is gas price and `Tm` is max fee per gas. I made the function to simplify that. I could even delete it cause it's only used twice and won't be used more times than that I guess.
<img width="740" alt="image" src="https://github.com/user-attachments/assets/36ff7184-fbd1-446a-a0ee-c232c24f8c9d">